lib/rules/no-dupe-disjunctions.ts

```diff
@@ -9,6 +10,492 @@ import type {
+                }
+            } else {
+                const lastIndex: number =
+                    md === "ltr" ? parent.elements.length - 1 : 0
+                if (parent.elements[lastIndex] !== p) {
+                    return false
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Returns whether the given RE AST contains assertions.
+ */
+function containsAssertions(expression: Expression): boolean {
+    try {
+        visitAst(expression, {
+            onAssertionEnter() {
+                throw new Error()
+            },
+        })
+        return false
+    } catch (error) {
+        return true
+    }
+}
+
+const creationOption: Transformers.CreationOptions = {
+    ignoreAmbiguity: true,
+    ignoreOrder: true,
+}
+const assertionTransformer = combineTransformers([
+    Transformers.applyAssertions(creationOption),
+    Transformers.removeUnnecessaryAssertions(creationOption),
+    Transformers.inline(creationOption),
+    Transformers.removeDeadBranches(creationOption),
+])
+
+/**
+ * Create an NFA from the given element.
+ */
+function toNFA(parser: JS.Parser, element: JS.ParsableElement): NFA {
+    try {
+        const { expression, maxCharacter } = parser.parseElement(element, {
+            backreferences: "disable",
+            assertions: "parse",
+        })
+
+        let e
+        if (containsAssertions(expression)) {
+            e = transform(assertionTransformer, expression)
+        } else {
+            e = expression
+        }
+
+        return NFA.fromRegex(e, { maxCharacter }, { assertions: "disable" })
+    } catch (error) {
+        return NFA.empty({
+            maxCharacter: parser.ast.flags.unicode ? 0x10ffff : 0xffff,
+        })
+    }
+}
+
+/**
+ * Unions all given NFAs
+ */
+function unionAll(nfas: readonly ReadonlyNFA[]): ReadonlyNFA {
+    if (nfas.length === 0) {
+        throw new Error("Cannot union 0 NFAs.")
+    } else if (nfas.length === 1) {
+        return nfas[0]
+    }
+
+    const total = nfas[0].copy()
+    for (let i = 1; i < nfas.length; i++) {
+        total.union(nfas[i])
+    }
+    return total
+}
+
+/**
+ * Returns whether one NFA is a subset of another.
+ */
+function isSubsetOf(
+    superset: ReadonlyNFA,
+    subset: ReadonlyNFA,
+): boolean | null {
+    try {
+        const a = DFA.fromIntersection(superset, subset)
+        const b = DFA.fromFA(subset)
+        a.minimize()
+        b.minimize()
+        return a.structurallyEqual(b)
+    } catch (error) {
+        return null
+    }
+}
+
+const enum SubsetRelation {
+    none,
+    leftEqualRight,
+    leftSubsetOfRight,
+    leftSupersetOfRight,
+    unknown,
+}
+
+/**
+ * Returns the subset relation
+ */
+function getSubsetRelation(
+    left: ReadonlyNFA,
+    right: ReadonlyNFA,
+): SubsetRelation {
+    try {
+        const inter = DFA.fromIntersection(left, right)
+        inter.minimize()
+
+        const l = DFA.fromFA(left)
+        l.minimize()
+
+        const r = DFA.fromFA(right)
```

Comment:
DFA constructions can blow up exponentially. Refa detects these exponential blowups and throws an error after a certain point. The default for this point is quite conservative at 10k nodes. For perspective, a DFA construction producing 10k nodes takes about 1 to 10ms, and one producing 100k nodes takes about 10 to 100ms, one producing 1M nodes takes about 100 to 1000ms, and so on.

DFA constructions are necessary to figure out which error message to use. If the DFA construction fails, we currently default to the "overlap" error message (instead of reporting e.g. a subset (=useless) alternative). Maybe we should increase the limit and potentially take more time for more accurate error messages?

Reply:
Does this mean there is a possibility of false positives?
I think false negatives are better than false positives, so if we get an error in DFA construction, I think we shouldn't report the error.
I think it is better if you can not limit the number of nodes.

Reply:
No, false positives aren't possible.

> I think it is better if you can not limit the number of nodes.

That's not a good idea. DFA construction blows up exponentially, so even a small regex can create a DFA that is gigabytes in size. I don't want your plugin crashing ESLint by eating up all the available memory :)

I'll increase the node limit to 100k.

Reply:
> No, false positives aren't possible.

Does that mean you may not be able to report an easy to understand message?
If so, I don't think there is any problem :+1:

Reply:
It means that we might report the wrong _type_ of message. 

I.e. for `A|B`, `B` might be a subset of `A` but if the DFA construction fails, we will report it as "overlap" instead of "subset".

There won't be any cases, where we tell the user that it's safe to remove something when it isn't. But there might be cases where something is safe to remove but we don't tell the user.